### PR TITLE
HTTP/3 QPACK - do not expect section ack for zero required insert count

### PR DESCRIPTION
**Pull request by lachlan-roberts**
https://datatracker.ietf.org/doc/html/draft-ietf-quic-qpack-21#section-4.4.1

> After processing an encoded field section whose declared Required
   Insert Count is not zero, the decoder emits a Section Acknowledgment
   instruction.

>  If an encoder receives a Section Acknowledgment instruction referring
   to a stream on which every encoded field section with a non-zero
   Required Insert Count has already been acknowledged, this MUST be
   treated as a connection error of type QPACK_DECODER_STREAM_ERROR.

### CODE REVIEW

**Review by sbordet — CHANGES_REQUESTED** [review]
Test case?

File: jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        instruction = _decoderHandler.getInstruction();
+        assertNull(instruction);
+    }
+
+    @Test
+    public void testSectionAcknowledgmentForZeroRequiredInsertCountOnEncoder() throws Exception
+    {
+        // Encode a header with only a value contained in the static table.
+        ByteBuffer buffer = encode(_encoder, 0, toMetaData("GET", "/", "http"));
+        System.err.println(BufferUtil.toDetailString(buffer));
```
Comment (sbordet) [inline jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java]: Remove the `System.err.`.

File: jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        // Parsing a section ack instruction on the encoder when we are not expecting it should result in QPACK_DECODER_STREAM_ERROR.
+        SectionAcknowledgmentInstruction instruction = new SectionAcknowledgmentInstruction(0);
+        ByteBuffer instructionBuffer = toBuffer(instruction);
+        QpackException error = assertThrows(QpackException.class, () -> _encoder.parseInstructions(instructionBuffer));
```
Comment (sbordet) [inline jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java]: You want to test for `QpackException.StreamException` here to be sure it's not a SESSION error.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java]: done

File: jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java (hunk 2 above)
Comment (lachlan-roberts) [reply jetty-http3/http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/SectionAcknowledgmentTest.java]: It is a `SessionException` and thats what I would expect it to be. The rfc says it must be treated "as a connection error".

I have added this check to the test case.

**Review by sbordet — APPROVED** [review]

**Pull request merged**